sql: fix event log related panic in NewSession

RaduBerinde opened the pull request:
Fixing an "event log under span" panic reported on Gitter. The session was
coming from `adminServer.GetUIData` (as part of `Server.maybeReportUsage`).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9975)

<!-- Reviewable:end -->

RaduBerinde committed:
sql: fix event log related panic in NewSession

Fixing an "event log under span" panic reported on Gitter. The session was
coming from `adminServer.GetUIData` (as part of `Server.maybeReportUsage`).

After some recent changes around AmbientContext, we have cases where we start a
SQL session under a span. This was not supported properly. We now create a
child span in this case (as opposed to a root span), and we don't require a
Tracer to be embedded in the context in this case.

Updating admin_test code to create a context with a span, like the real server
code. With these changes, the tests reproduce the crash.

andreimatei commented:
LGTM

can/should you file a bug for whomever knows about the adminServer (maybe @mrtracy) to add a test exercising this?

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/session.go, line 169 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-KU2cBpuc9xmfv1uHf96:-KU2c6GgklX3nHXPba9D:blhu2br) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d88994ac4e5de8399d99f4cd2389c69db8d184a/pkg/sql/session.go#L169)):*
> ```Go
> 	if opentracing.SpanFromContext(ctx) == nil {
> 		// Set up an EventLog for session events.
> 		ctx = log.WithEventLog(ctx, fmt.Sprintf("sql [%s]", args.User), remoteStr)
> ```

hmm if Sessions have an event log, and we do the thing where a user can enable/disable tracing on a session (which is gonna be a span), then that trace will make the session stop reporting to the event log. So maybe it'll be good to have control over whether logs go to both the eventlog and the span or not (either through hiding the event log behind a special tracer and using a tee or otherwise).

---

*[pkg/sql/session.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-KU2d9TP2C_SdYFXIx8K:-KU2d9TP2C_SdYFXIx8L:b-op7hup) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d88994ac4e5de8399d99f4cd2389c69db8d184a/pkg/sql/session.go#L184)):*
> ```Go
> 	// addressed, there might be leases accumulated by preparing statements.
> 	s.planner.releaseLeases()
> 	if s.eventLogInCtx {
> ```

can't we just inspect the context? Or is the question whether the Session owns the `eventLog`? If so, maybe rename the bool accordingly. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-:-KU2dNBG9ByF9nz06sU0:bmhirnl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
It looks like the start of the operation is in server code,

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/session.go, line 169 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-KU2cBpuc9xmfv1uHf96:-KU2m77MHeNHuwSpIFRI:b-bi0wcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d88994ac4e5de8399d99f4cd2389c69db8d184a/pkg/sql/session.go#L169)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> hmm if Sessions have an event log, and we do the thing where a user can enable/disable tracing on a session (which is gonna be a span), then that trace will make the session stop reporting to the event log. So maybe it'll be good to have control over whether logs go to both the eventlog and the span or not (either through hiding the event log behind a special tracer and using a tee or otherwise).
</details>
The fact that we can get an actual context also puts a dent into our thinking yesterday that Session would have an AmbientCtx instead of a Context

---

*[pkg/sql/session.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-KU2d9TP2C_SdYFXIx8K:-KU2lrAkIZ1Z11uXXZL2:b99a7jq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d88994ac4e5de8399d99f4cd2389c69db8d184a/pkg/sql/session.go#L184)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can't we just inspect the context? Or is the question whether the Session owns the `eventLog`? If so, maybe rename the bool accordingly. 
</details>
We can call `FinishEventLog` unconditionally, but my thinking was if the existing context already has a different event log. Not that we do ever do that right now. I'll rename the variable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-:-KU2ldkFIEo4VmVupbSm:byjj132)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I'll look into adding a test

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-:-KU2mYLPX00u4LQEFxNr:b-qugnak)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Updated. The admin_test code now exercises this fix. I fixed a couple of other things related to starting a SQL session under an existing span.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-:-KU7d8ZDxA-9A8IBqKA8:bg8s9fz)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
still LGTM

merge this when you can cause I'm also poking around sessuib tracing

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/session.go, line 169 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-KU2cBpuc9xmfv1uHf96:-KU7tp0tPTE0F8GSk6Ny:b9950x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d88994ac4e5de8399d99f4cd2389c69db8d184a/pkg/sql/session.go#L169)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The fact that we can get an actual context also puts a dent into our thinking yesterday that Session would have an AmbientCtx instead of a Context
</details>
indeed. I guess we want to leave the Session having a Context.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9975#-:-KU7sKPMQXnMlb3Sjn0z:b-kyvr2m)*
<!-- Sent from Reviewable.io -->
